daemon/list.go

```diff
@@ -23,17 +24,33 @@ type iterationAction int
 // Returns the object to serialize by the api.
 type containerReducer func(*Container, *listContext) (*types.Container, error)
 
-const (
-	// includeContainer is the action to include a container in the reducer.
-	includeContainer iterationAction = iota
-	// excludeContainer is the action to exclude a container in the reducer.
-	excludeContainer
-	// stopIteration is the action to stop iterating over the list of containers.
-	stopIteration
-)
+// containerList represent a list to store containers.
+// This struct is concurrently safe.
+type containerList struct {
+	list []*types.Container
+	m    sync.Mutex
+}
+
+// newContainerList creates a new list of containers with a predefined capacity.
+func newContainerList(capacity int) *containerList {
+	return &containerList{
+		list: make([]*types.Container, 0, capacity),
+	}
+}
 
-// errStopIteration makes the iterator to stop without returning an error.
-var errStopIteration = errors.New("container list iteration stopped")
+// Add appends containers to the list.
+func (c *containerList) Add(container *types.Container) {
+	c.m.Lock()
+	c.list = append(c.list, container)
+	c.m.Unlock()
+}
+
+// List returns the collection of containers in a safe way.
+func (c *containerList) List() []*types.Container {
+	c.m.Lock()
```

Comment:
Not sure that I understand motivation of locking here, does it really can be called concurrently with `Add`? And in this case returning ref is unsafe, because calling code can mess with it.